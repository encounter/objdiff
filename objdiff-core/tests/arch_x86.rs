use objdiff_core::{diff, diff::display::SymbolFilter, obj};

mod common;

#[test]
#[cfg(feature = "x86")]
fn read_x86() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86/staticdebug.obj"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx = obj.symbols.iter().position(|s| s.name == "?PrintThing@@YAXXZ").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}

#[test]
#[cfg(feature = "x86")]
fn read_x86_combine_sections() {
    let diff_config = diff::DiffObjConfig {
        combine_data_sections: true,
        combine_text_sections: true,
        ..Default::default()
    };
    let obj = obj::read::parse(include_object!("data/x86/rtest.obj"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj.sections);
}

#[test]
#[cfg(feature = "x86")]
fn read_x86_64() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86_64/vs2022.o"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx =
        obj.symbols.iter().position(|s| s.name == "?Dot@Vector@@QEAAMPEAU1@@Z").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}

#[test]
#[cfg(feature = "x86")]
fn display_section_ordering() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86/basenode.obj"), &diff_config, obj::DiffSide::Base).unwrap();
    let obj_diff =
        diff::diff_objs(Some(&obj), None, None, &diff_config, &diff::MappingConfig::default())
            .unwrap()
            .left
            .unwrap();
    let section_display =
        diff::display::display_sections(&obj, &obj_diff, SymbolFilter::None, false, false, false);
    insta::assert_debug_snapshot!(section_display);
}

#[test]
#[cfg(feature = "x86")]
fn read_x86_jumptable() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86/jumptable.o"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx = obj.symbols.iter().position(|s| s.name == "?test@@YAHH@Z").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}

// Inferred size of functions should ignore symbols with specific prefixes
#[test]
#[cfg(feature = "x86")]
fn read_x86_local_labels() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/x86/local_labels.obj"), &diff_config, obj::DiffSide::Base).unwrap();
    insta::assert_debug_snapshot!(obj);
}
