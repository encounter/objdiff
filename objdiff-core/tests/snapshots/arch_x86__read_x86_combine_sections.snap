---
source: objdiff-core/tests/arch_x86.rs
expression: obj.sections
---
[
    Section {
        id: ".drectve-0",
        name: ".drectve",
        address: 0,
        size: 47,
        kind: Unknown,
        data: SectionData(
            0,
        ),
        flags: FlagSet(),
        align: Some(
            1,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".debug$S-0",
        name: ".debug$S",
        address: 0,
        size: 100,
        kind: Unknown,
        data: SectionData(
            0,
        ),
        flags: FlagSet(),
        align: Some(
            1,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".rdata-0",
        name: ".rdata",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".rdata-1",
        name: ".rdata",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".text$mn-0",
        name: ".text$mn",
        address: 0,
        size: 0,
        kind: Code,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            16,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".data-combined",
        name: ".data",
        address: 0,
        size: 56,
        kind: Data,
        data: SectionData(
            56,
        ),
        flags: FlagSet(Combined),
        align: Some(
            4,
        ),
        relocations: [
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 0,
                target_symbol: 44,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 16,
                target_symbol: 44,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 32,
                target_symbol: 44,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 48,
                target_symbol: 6,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 52,
                target_symbol: 8,
                addend: 0,
            },
        ],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".rdata-combined",
        name: ".rdata",
        address: 0,
        size: 304,
        kind: Data,
        data: SectionData(
            304,
        ),
        flags: FlagSet(Combined),
        align: Some(
            4,
        ),
        relocations: [
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 12,
                target_symbol: 17,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 16,
                target_symbol: 19,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 24,
                target_symbol: 13,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 48,
                target_symbol: 15,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 64,
                target_symbol: 25,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 68,
                target_symbol: 27,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 76,
                target_symbol: 21,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 100,
                target_symbol: 23,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 116,
                target_symbol: 31,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 120,
                target_symbol: 33,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 136,
                target_symbol: 35,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 140,
                target_symbol: 37,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 144,
                target_symbol: 19,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 148,
                target_symbol: 39,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 156,
                target_symbol: 31,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 180,
                target_symbol: 33,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 184,
                target_symbol: 21,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 208,
                target_symbol: 23,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 224,
                target_symbol: 31,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 228,
                target_symbol: 33,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 244,
                target_symbol: 13,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 248,
                target_symbol: 15,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 264,
                target_symbol: 21,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 268,
                target_symbol: 23,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 272,
                target_symbol: 29,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 276,
                target_symbol: 11,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 280,
                target_symbol: 43,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 284,
                target_symbol: 41,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 288,
                target_symbol: 70,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 292,
                target_symbol: 56,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 296,
                target_symbol: 72,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 300,
                target_symbol: 59,
                addend: 0,
            },
        ],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".rdata$r-1",
        name: ".rdata$r",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".rdata$r-2",
        name: ".rdata$r",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".data$rs-1",
        name: ".data$rs",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".rdata$r-3",
        name: ".rdata$r",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".rdata$r-4",
        name: ".rdata$r",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".rdata$r-5",
        name: ".rdata$r",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".rdata$r-6",
        name: ".rdata$r",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".data$rs-2",
        name: ".data$rs",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".rdata$r-7",
        name: ".rdata$r",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".rdata$r-8",
        name: ".rdata$r",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".rdata$r-9",
        name: ".rdata$r",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".rdata$r-10",
        name: ".rdata$r",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".text$mn-1",
        name: ".text$mn",
        address: 0,
        size: 0,
        kind: Code,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            1,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".rdata$r-11",
        name: ".rdata$r",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".text$mn-2",
        name: ".text$mn",
        address: 0,
        size: 0,
        kind: Code,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            16,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".text$mn-3",
        name: ".text$mn",
        address: 0,
        size: 0,
        kind: Code,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            16,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".text$mn-4",
        name: ".text$mn",
        address: 0,
        size: 0,
        kind: Code,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            16,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".text$mn-5",
        name: ".text$mn",
        address: 0,
        size: 0,
        kind: Code,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            16,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".text$mn-6",
        name: ".text$mn",
        address: 0,
        size: 0,
        kind: Code,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            16,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".text-combined",
        name: ".text",
        address: 0,
        size: 320,
        kind: Code,
        data: SectionData(
            320,
        ),
        flags: FlagSet(Combined),
        align: Some(
            16,
        ),
        relocations: [
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 4,
                target_symbol: 62,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    20,
                ),
                address: 9,
                target_symbol: 53,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    20,
                ),
                address: 43,
                target_symbol: 60,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    20,
                ),
                address: 62,
                target_symbol: 52,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    20,
                ),
                address: 84,
                target_symbol: 11,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 100,
                target_symbol: 64,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 124,
                target_symbol: 66,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 156,
                target_symbol: 6,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 166,
                target_symbol: 8,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    20,
                ),
                address: 177,
                target_symbol: 57,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    20,
                ),
                address: 185,
                target_symbol: 54,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    20,
                ),
                address: 219,
                target_symbol: 54,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    20,
                ),
                address: 238,
                target_symbol: 52,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    20,
                ),
                address: 267,
                target_symbol: 57,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    20,
                ),
                address: 286,
                target_symbol: 52,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 308,
                target_symbol: 68,
                addend: 0,
            },
            Relocation {
                flags: Coff(
                    20,
                ),
                address: 313,
                target_symbol: 60,
                addend: 0,
            },
        ],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".text$yd-0",
        name: ".text$yd",
        address: 0,
        size: 0,
        kind: Code,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            16,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".rdata-2",
        name: ".rdata",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".rdata-3",
        name: ".rdata",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".data-0",
        name: ".data",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".rdata$r-12",
        name: ".rdata$r",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".rdata$r-13",
        name: ".rdata$r",
        address: 0,
        size: 0,
        kind: Data,
        data: SectionData(
            0,
        ),
        flags: FlagSet(Hidden),
        align: Some(
            4,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".CRT$XCU-0",
        name: ".CRT$XCU",
        address: 0,
        size: 4,
        kind: Data,
        data: SectionData(
            4,
        ),
        flags: FlagSet(),
        align: Some(
            4,
        ),
        relocations: [
            Relocation {
                flags: Coff(
                    6,
                ),
                address: 0,
                target_symbol: 61,
                addend: 0,
            },
        ],
        line_info: {},
        virtual_address: None,
    },
    Section {
        id: ".chks64-0",
        name: ".chks64",
        address: 0,
        size: 280,
        kind: Unknown,
        data: SectionData(
            0,
        ),
        flags: FlagSet(),
        align: Some(
            16,
        ),
        relocations: [],
        line_info: {},
        virtual_address: None,
    },
]
